pravega/pravega-client-rust
# Issue 357: Nodejs ReaderGroup and EventReader

Signed-off-by: thekingofcity <3353040+thekingofcity@users.noreply.github.com>

**Change log description**  
Add `ReaderGroup` and `EventReader` support to Nodejs bindings.

**Purpose of the change**  
Fixes #357 

**What the code does**  
Implement the `stream_reader_group` and `stream_reader` just like the Python code.

It also uses the asynchronous approach so that users can do the followings:
```javascript
const seg_slice: Slice = await stream_reader.get_segment_slice();
const enc = new TextDecoder('utf-8');
for (const event of seg_slice) {
    const raw_value: ArrayBuffer = event.data();
    console.log(`Event at ${event.offset()} reads ${enc.decode(raw_value)}`);
}
```

This should be merged after #355.

Reference about `@@iterator`:
1. https://www.geekabyte.io/2019/06/typing-iterables-and-iterators-with.html
2. https://basarat.gitbook.io/typescript/future-javascript/iterators
3. https://stackoverflow.com/questions/52326671/the-syntax-of-symbol-iterator-is-difficult
4. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols


**How to verify it**  
Since the writer is not here, I use Python code to create the stream and write something to Pravega.

1. `npm run build-debug`
2. `python3 tests/write.py` will write 4 events.
  ```
  b'e1'
  b'e2'
  b'e3'
  b'e4'
  ```
3. `node --loader ts-node/esm tests/stream_reader.ts` should print these.
  ```
  (node:5239) ExperimentalWarning: --experimental-loader is an experimental feature. This feature could change at any time
  (Use `node --trace-warnings ...` to show where the warning was created)
  Event at 0 reads e1
  Event at 10 reads e2
  Event at 20 reads e3
  Event at 30 reads e4
  ```

I will write similar tests like Python after the writer is implemented.
